File: luasrc/model/cbi/change-mac.lua
-- Copyright 2020 muink <https://github.com/muink>
-- Licensed to the public under the MIT License

local uci	= require "luci.model.uci".cursor()
local fs	= require("nixio.fs")
local sys	= require("luci.sys")
local util	= require("luci.util")
local packageName = "change-mac"
local conf = packageName
local typeds = ".@change-mac[0]."
local config = "/etc/config/" .. conf


m = Map(conf, translate("MAC address randomizer"), translate("Assign a random MAC address to the designated interface on every time boot"))
m.pageaction = false



s = m:section(TypedSection, "change-mac")
s.anonymous = true

enabled = s:option(Flag, "enabled", translate("Enable MAC randomization"))
enabled.rmempty = false

network = s:option(DynamicList, "network", translate("Enabled interfaces"))
network.template = "cbi/network_netlist"
network.novirtual = true --not supported VLAN
network.rmempty = false
network.widget = "checkbox"

merge_physical = s:option(Flag, "merge_physical", translate("Merge the same physical interface"))
merge_physical.rmempty = false

restore_sel = s:option(Button, "_restore_sel", translate("Restore selected interfaces"))
restore_sel.inputtitle = translate("Restore selected interfaces")
restore_sel.inputstyle = "apply"
restore_sel.write = function()
	--m.uci:save(conf)
	m.uci:commit(conf)
	m.uci:apply()

	sys.call ("/etc/init.d/change-mac restore")
end


random_mode = s:option(ListValue, "random_mode", translate("Multi-interface MAC random mode"))
random_mode:value("disorderly", translate("Disorderly"))
random_mode:value("sequence", translate("Sequence"))
random_mode.default = "disorderly"
random_mode.rmempty = false

mac_type = s:option(ListValue, "mac_type", translate("MAC address type"),
	translate("Use command 'rgmac --help' to get more information"))
mac_type:value("locally", translate("Locally administered address"))
mac_type:value("specific", translate("Specify OUI"))
mac_type:value("vendor", translate("Vendor name"))
mac_type.default = "locally"
mac_type.rmempty = false

mac_type_specific = s:option(Value, "mac_type_specific", translate("Specify OUI"))
mac_type_specific.placeholder = "74:D0:2B"
--mac_type_specific:depends("mac_type", "specific")
mac_type_specific.rmempty = false

mac_type_vendor = s:option(Value, "mac_type_vendor", translate("Vendor name"),
	translate("Use command 'rgmac -lrouter' to get valid vendor name"))
mac_type_vendor.placeholder = "router:Asus"
--mac_type_vendor:depends("mac_type", "vendor")
mac_type_vendor.rmempty = false

change_now = s:option(Button, "_change_now", translate("Change MAC now"))
change_now.inputtitle = translate("Change MAC now")
change_now.inputstyle = "apply"
change_now.write = function()
	--m.uci:save(conf)
	m.uci:commit(conf)
	m.uci:apply()

	sys.call ("/etc/init.d/change-mac start")
end

save_apply = s:option(Button, "_save_apply", translate("Save & Apply"))
save_apply.inputtitle = translate("Save & Apply")
save_apply.inputstyle = "apply"
save_apply.write = function()
	--m.uci:save(conf)
	m.uci:commit(conf)
	m.uci:apply()

	if tostring(util.trim(sys.exec("uci get " .. conf .. typeds .. "enabled"))) == "1" then sys.call ("/etc/init.d/change-mac enable");
	else sys.call ("/etc/init.d/change-mac disable"); end
end


return m

File: htdocs/luci-static/resources/view/change-mac.js
'use strict';
'require view';
'require fs';
'require uci';
'require ui';
'require form';
'require tools.widgets as widgets';

return view.extend({
//	handleSaveApply: null,
//	handleSave: null,
//	handleReset: null,

//	callInitAction: rpc.declare({
//		object: 'luci',
//		method: 'setInitAction',
//		params: [ 'name', 'action' ],
//		expect: { result: false }
//	}),

	load: function() {
	return Promise.all([
		uci.load('network'),
		uci.load('change-mac'),
	]);
	},

//	handleAction: function(name, action, ev) {
//		return this.callInitAction(name, action).then(function(success) {
//			if (success != true)
//				throw _('Command failed');
//
//			return true;
//		}).catch(function(e) {
//			ui.addNotification(null, E('p', _('Failed to execute "/etc/init.d/%s %s" action: %s').format(name, action, e)));
//		});
//	},

	handleAction: function(m, action, ev) {
		m.save();
		uci.save();
		uci.apply();
		uci.unload('change-mac');
		uci.load('change-mac');

		return fs.exec('/etc/init.d/change-mac', [action])
			.then(L.bind(uci.unload, uci, 'change-mac'))
			.then(L.bind(m.render, m))
			.catch(function(e) { ui.addNotification(null, E('p', e.message)) });
	},

	render: function() {
		var m, s, o;

		m = new form.Map('change-mac', _('MAC address randomizer'),
			_('Assign a random MAC address to the designated interface on every time boot'));

		s = m.section(form.TypedSection, 'change-mac');
		s.anonymous = true;

		o = s.option(form.Flag, 'enabled', _('Enable MAC randomization'));
		o.rmempty = false;

		o = s.option(widgets.DeviceSelect, 'interface', _('Enabled interfaces'));
		o.multiple = true;
		o.noaliases = true;
		o.nobridges = true;
		o.nocreate = true;

		o = s.option(form.ListValue, 'random_mode', _('Multi-interface MAC random mode'));
		o.value('disorderly', _('Disorderly'));
		o.value('sequence', _('Sequence'));
		o.default = 'disorderly';
		o.rmempty = false;

		o = s.option(form.ListValue, 'mac_type', _('MAC address type'),
			_("Use command 'rgmac --help' to get more information"));
		o.value('locally', _('Locally administered address'));
		o.value('specific', _('Specify OUI'));
		o.value('vendor', _('Vendor name'));
		o.default = 'locally';
		o.rmempty = false;

		o = s.option(form.Value, 'mac_type_specific', _('Specify OUI'));
		o.placeholder = '74:D0:2B';
		//o.depends('mac_type', 'specific');
		o.rmempty = false;

		o = s.option(form.Value, 'mac_type_vendor', _('Vendor name'),
			_("Use command 'rgmac -lrouter' to get valid vendor name"));
		o.placeholder = 'router:Asus';
		//o.depends('mac_type', 'vendor');
		o.rmempty = false;

		o = s.option(form.Button, '_change_now', _('Change MAC now'));
		o.inputtitle = _('Change now');
		o.inputstyle = 'apply';
		o.onclick = this.handleAction.bind(this, m, 'start');
// E('button', { 'class': 'btn cbi-button-action', 'click': ui.createHandlerFn(this, 'handleAction', list[i].name, 'start'), 'disabled': isReadonlyView }, _('Start')),

		o = s.option(form.Button, '_restore_sel', _('Restore selected interfaces'));
		o.inputtitle = _('Restore');
		o.inputstyle = 'apply';
		o.onclick = this.handleAction.bind(this, m, 'restore');
// E('button', { 'class': 'btn cbi-button-action', 'click': ui.createHandlerFn(this, 'handleAction', list[i].name, 'stop'), 'disabled': isReadonlyView }, _('Stop')),

		o = s.option(form.Button, '_query_oui', _('Query OUI vendor'));
		o.inputtitle = _('Query');
		o.inputstyle = 'apply';
		o.onclick = function() {
		}

		return m.render();
	}
});
